### PR TITLE
Fix PyTorch Object Detection Estimators Missing Gradients Bug

### LINKED ISSUE #2248: DPatch Attack Error in `PyTorchObjectDetector`
**Describe the bug**
When running the DPatch attack for the `PyTorchObjectDetector` and `PyTorchFasterRCNN`, an error can occur when using any PyTorch-specific preprocessors.

**To Reproduce**
Run the DPatch patch attack on the `PyTorchObjectDetector` while using a Faster R-CNN model. When using additional PyTorch-specific preprocessors, the error above will occur.

**Expected behavior**
The attacks should run without error. It can be easily fixed by adding the following line to the `_get_losses` method in the `PyTorchObjectDetector` class:

```python
if x_preprocessed.is_leaf:
    x_preprocessed.requires_grad = True
else:
    x_preprocessed.retain_grad()
```

This error occurs because the gradient retention property is lost when moving a tensor to a different device.

**Screenshots**

**System information (please complete the following information):**
 - OS
 - Python version
 - ART version or commit number
 - TensorFlow / Keras / PyTorch / MXNet version

### PR DESCRIPTION
**Pull request by f4str**
# Description

When a non-leaf tensor is moved to another device, the gradient retention property is lost and needs to be manually set again. This PR adds this minor change to both the `PyTorchObjectDetector` and `PyTorchYolo` estimators.

Fixes #2248

## Type of change

Please check all relevant options.

- [ ] Improvement (non-breaking)
- [x] Bug fix (non-breaking)
- [ ] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Testing

Please describe the tests that you ran to verify your changes. Consider listing any relevant details of your test configuration.

- [x] Tests for `PyTorchObjectDetector` are unchanged
- [x] Tests for `PyTorchFasterRCNN` are unchanged

**Test Configuration**:
- OS
- Python version
- ART version or commit number
- TensorFlow / Keras / PyTorch / MXNet version

# Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by beat-buesser** [comment]
Hi @f4str Thank you for your pull request. In #2238 we made the same change also in method `PyTorchObjectDetector._preprocess_and_convert_inputs`. Is the same change also needed in `PyTorchYolo._preprocess_and_convert_inputs`?

**Comment by f4str** [comment]
Hi @beat-buesser this is a separate bug-fix independent from #2238 that affects both the `PyTorchObjectDetector` and `PyTorchYolo`. The previous change addressed a different bug that only occured for the `PyTorchObjectDetector`.

**Review by beat-buesser — APPROVED** [review]
Hi @f4str Thank you very much for noticing and fixing this bug!
